### PR TITLE
Adds a new Banning Panel made with TGUI

### PR DESCRIPTION
**Pull request by Miliviu**
## About The Pull Request

Adds a new modern and cool Banning Panel done with TGUI

## Why It's Good For The Game

QOL for the Admin team, it's pretty and easy to use, and if anyone doesn't like the new panel (or TGUI dies) the old one is still available.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

https://github.com/BeeStation/BeeStation-Hornet/assets/45698448/a6129592-e55e-49ca-b81a-2e26f8b401e0

</details>

## Changelog
:cl: jixs4v, Miliviu, Ratón
admin: Added a new Banning Panel created with TGUI, old panel is still available with the verb old-banning-panel
/:cl:

### CODE REVIEW

**Review by EvilDragonfiend — DISMISSED** [review]
The special global bans like `BAN_ROLE_ALL_ANTAGONISTS` should be accessible without opening the list. I don't like our bans become bloated when there is a ban option for literally all antagonists.

**Comment by Miliviu** [comment]
Now :>

**Review by JixS4v — CHANGES_REQUESTED** [review]
Needs force cryo after

https://github.com/Miliviu/BeeStation-Hornet/pull/1

**Comment by JixS4v** [comment]
Taking a look at the full PR body, will come back with a more detailed review

**Review by JixS4v — APPROVED** [review]
LGTM after testing!

**Review by itsmeow — CHANGES_REQUESTED** [review]
Some layout feedback:

Move ban reason into its own section. It should be the first thing after Key/IP/CID
Move Duration, Type, Applies to Admins, suppression, and Force cryo after into its own section, after the reason.

The duration input should have a permanent checkbox that hides the time input that is off by default, rather than a dropdown to switch types.

Enabling Key, IP, etc should create input fields on a new line below, with a label correspond to what it is

```
- [X] IP
- [X] CID

...

IP: [INPUT HERE]
CID: [INPUT HERE]
```

"SUBMIT" Should be a big green fucking button, next to an equally big red CANCEL. It should be at the bottom of the banning panel. See TGUI inputs for this.

Role bans should use multiple columns if there is horizontal space to do so (you can use a flex with a 50% basis to do this)
Flip the order of the role ban category buttons. Pressing the name should expand it. There should be a color='transparent' checkbox on the right that expands and selects all for a category.

There should be some way to search for roles that filters the shown roles and expands all categories temporarily while searching.

If possible, move abstract roles to the top and make them always expanded rather than their own category. Mainly because Appearance bans are very common.

(No code review atm)

File: code/modules/admin/sql_ban_system.dm [hunk 1]
```diff
@@ -1094,3 +1077,233 @@
...
+							<div class='inputbox[(job in banned_from) ? " banned" : ""]'></div></label>
+				"}
+				break_counter++
+			output += "</div></div>"
+		var/list/long_job_lists = list(
+			"Civilian" = GLOB.civilian_positions | JOB_NAME_GIMMICK,
+			"Antagonist Positions" = list(BAN_ROLE_ALL_ANTAGONISTS) + GLOB.antagonist_bannable_roles,
+			"Forced Antagonist Positions" = list(BAN_ROLE_FORCED_ANTAGONISTS) + GLOB.forced_bannable_roles,
+			"Ghost Roles" = list(BAN_ROLE_ALL_GHOST) + GLOB.ghost_role_bannable_roles,
+			"Other" = GLOB.other_bannable_roles,
```
Comment (itsmeow) [inline code/modules/admin/sql_ban_system.dm:1291]: We really, really should not have this in two places (it's also a static list on the tgui datum)

These things should be put in a global variable or something.

**Comment by JixS4v** [comment]
Hello, admin (joelogbybolb) and original designer of the UI that was behind this here, with miliviu picking up where I left and finishing it (thankfully)

image
Here's a quick mockup of the proposed designed decisions (save the search bar because it's a bit less easy). I'll get into why I don't agree with some of these changes.
> Move ban reason into its own section. It should be the first thing after Key/IP/CID 

This is plain not needed, the ban reasons rarely exceed the space that was alloted in the original UI and it's just adding bulk to the window without much substance.
> Move Duration, Type, Applies to Admins, suppression, and Force cryo after into its own section, after the reason.

At risk of pulling the "legacy" card, this is how it was organized before so admins won't have to adjust to the change, furthermore the space alloted is plenty and we usually set these before actually writing the ban so it's not really needed.

> The duration input should have a permanent checkbox that hides the time input that is off by default, rather than a dropdown to switch types.

This makes the UI much less clear and intuitive to use, and usually temporary bans are the default anyways.

> Enabling Key, IP, etc should create input fields on a new line below, with a label correspond to what it is
> 
> ```
> - [X] IP
> - [X] CID
> 
> ...
> 
> IP: [INPUT HERE]
> CID: [INPUT HERE]
> ```
> 
I don't think this is a positive change, it spreads out the UI for no reason?
> "SUBMIT" Should be a big green fucking button, next to an equally big red CANCEL. It should be at the bottom of the banning panel. See TGUI inputs for this.

Why would you need a cancel button? This isn't a prompt, it's a panel you open and you will usually make several bans from the same panel instance, for instance when applying a combined server+role ban
> Role bans should use multiple columns if there is horizontal space to do so (you can use a flex with a 50% basis to do this)

This looks nicer! My original intention behind the single-column layout is that finding categories is much easier, since you can just hover your mouse while scrolling till you find the right one.

> Flip the order of the role ban category buttons. Pressing the name should expand it. There should be a color='transparent' checkbox on the right that expands and selects all for a category.

This is purely aesthetic but I do think it looks nicer the other way.

> There should be some way to search for roles that filters the shown roles and expands all categories temporarily while searching.

There's so few roles to ban and you're usually banning general categories anyways (as per admin policy!), so adding a search bar is pretty unecessary

> If possible, move abstract roles to the top and make them always expanded rather than their own category. Mainly because Appearance bans are very common.

Appearance bans are not common at all, I have only appearance banned someone once in my almost 5 months of adminning

All this being said, I will hold a poll with the rest of the admin team to see what they prefer.

**Comment by itsmeow** [comment]
> I don't think this is a positive change, it spreads out the UI for no reason?

It's designed to reduce layout shift, where the UI moves buttons when you click something. It makes UI more difficult to use.

Alternatively, you could ensure each toggleable input is on a new line next to its checkbox and label.


Also, I think you misunderstood my duration request: It should be a checkbox labeled "Permanent" that is OFF by default. If it is off, it will show the duration input. So temp ban is the default.

**Comment by itsmeow** [comment]
Also, for the Ban Details tab, you can move that into the top section as another column using flex.

**Comment by itsmeow** [comment]
For role ban sections, I'll agree with you, the old one is better, although maybe try setting the icon position to right?

**Comment by JixS4v** [comment]
> It's designed to reduce layout shift, where the UI moves buttons when you click something. It makes UI more difficult to use.

would there be a way to make it gray the input out instead of just making it dissapear? that'd be ideal, in my opinion.

**Comment by JixS4v** [comment]
> For role ban sections, I'll agree with you, the old one is better, although maybe try setting the icon position to right?

The checkbox position you mean? I think it's fine to the left because otherwise the collapsible opening shifts it to the end of the contents.

**Comment by itsmeow** [comment]
No, the arrow

**Comment by JixS4v** [comment]
> No, the arrow

Don't think it's possible without changing the component

**Comment by itsmeow** [comment]
> 
> > It's designed to reduce layout shift, where the UI moves buttons when you click something. It makes UI more difficult to use.
> 
> would there be a way to make it gray the input out instead of just making it dissapear? that'd be ideal, in my opinion.
> 

Yes, but the second goal is to reduce visual clutter, which the old one is chock full of.

Just do this;
Key [X] [______]
IP [ ]
CID [ ]

**Comment by itsmeow** [comment]
The component already has an iconPosition prop.

**Comment by JixS4v** [comment]
I still don't think permanent bans being a checkbox makes little sense, but I'll get to implementing this tomorrow, it's pretty late.

**Comment by PowerfulBacon** [comment]
The old admin panel should just be removed. There is no reason to keep both, TGUI failing is uncommon and will result in the admin tools being unusable anyway

**Comment by JixS4v** [comment]
> > > It's designed to reduce layout shift, where the UI moves buttons when you click something. It makes UI more difficult to use.
> > 
> > 
> > would there be a way to make it gray the input out instead of just making it dissapear? that'd be ideal, in my opinion.
> 
> Yes, but the second goal is to reduce visual clutter, which the old one is chock full of.
> 
> Just do this; Key [X] [______] IP [ ] CID [ ]

image

I think this is ideal

File: code/modules/admin/sql_ban_system.dm (hunk 1 above)
Comment (Miliviu) [reply code/modules/admin/sql_ban_system.dm:1291]: Okay, I've checked this out, this kinda is in two places but not in the same way

Once in the tgui datum as a const list, I don't mind moving it into a glob variable it'd be fine, but this is in the OLD banning panel, I didn't plan touching it at all, I just changed the name to old_banning_panel...

It also doesn't have one list with all the jobs like our banning panel does, it has multiple lists with different jobs and stuff...

I can try to modify the old banning pannel but i'd honestly prefer not doing that!

**Comment by itsmeow** [comment]
Use a NumberInput for the duration

File: code/modules/admin/holder2.dm [hunk 2]
```diff
@@ -36,6 +36,9 @@ GLOBAL_PROTECT(href_token)
 	/// Player panel
 	var/datum/admin_player_panel/player_panel
 
+	/// Bannig Panel
```
Comment (JixS4v) [inline code/modules/admin/holder2.dm]:
```suggestion
	/// Banning Panel
```

**Comment by JixS4v** [comment]
> 🆑 jixs4v, Miliviu
Forgot ratón who helped out with a lot of the sql stuff.

**Comment by PowerfulBacon** [comment]
Those checkboxes should really have no background, since they go green when they are checked which is confusing when the background is green to begin with

**Comment by Miliviu** [comment]
@PowerfulBacon ^

File: tgui/packages/tgui/interfaces/BanningPanel.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                <Button.Checkbox content="Force Cryo After" checked={force_cryo_after} onClick={() => act('toggle_cryo')} />
+                <Button.Checkbox
+                  content="Applies to admins"
+                  checked={applies_to_admins}
+                  onClick={() => act('toggle_applies_to_admins')}
+                />
+
+                <Box style={can_supress ? { display: 'flex' } : { display: 'none' }}>
+                  <Button.Checkbox
+                    content="Enable supression"
```
Comment (itsmeow) [inline tgui/packages/tgui/interfaces/BanningPanel.js]:
```suggestion
                    content="Enable suppression"
```

**Review by itsmeow — CHANGES_REQUESTED** [review]
typo

File: tgui/packages/tgui/interfaces/BanningPanel.js [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+      <Window.Content>
+        <Section title="Player Information">
+          <Stack>
+            <Flex direction="column">
+              <Flex.Item>
+                <Button.Checkbox content="Key" checked={key_enabled} onClick={() => act('toggle_key')} minWidth="55px" />
+                <Input
+                  value={key}
+                  onChange={(e) => act('update_key', { key: e.target.value })}
+                  style={key_enabled ? { display: 'inline-block' } : { display: 'none' }}
```
Comment (itsmeow) [inline tgui/packages/tgui/interfaces/BanningPanel.js]: add `fluid` to these so they grow to fill space. also adding a placeholder attribute makes it easier to parse at first glance

File: tgui/packages/tgui/interfaces/BanningPanel.js (hunk 3 above)
Comment (Miliviu) [reply tgui/packages/tgui/interfaces/BanningPanel.js]: Thanks!

**Comment by Miliviu** [comment]
image

**Comment by Miliviu** [comment]
image

New version

**Review by itsmeow — APPROVED** [review]
New one is good but imo remove the old panel

**Review by Rukofamicom — APPROVED** [review]
Removing the old panel should happen, but I'm alright with that being done in a future PR so admins have time to adapt.

**Comment by Miliviu** [comment]
I commit to remove it in a future PR then!

**Pull request merged**